# 1. Select SRR file to analyse
PRJNA681021 \
[SRA Run Selector link](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_6644aabf3aadf42eac8fc81c&o=acc_s%3Aa&s=SRR13199080,SRR13199081,SRR13199085,SRR13199087,SRR13199088,SRR13199089)

> For normal people sample\
> SRR13199080 \
> SRR13199081 \
> SRR13199085

> For HIV-infected people sample \
> SRR13199087 \
> SRR13199088 \
> SRR13199089

# 2. Prefetch the SRR files
https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump

```sh
$ prefetch SRR13199080 --max-size 10t SRR13199081 --max-size 10t SRR13199085 --max-size 10t
```

```sh
$ prefetch SRR13199087 --max-size 10t SRR13199088 --max-size 10t SRR13199089 --max-size 10t
```


# 3. Making Fastq files from the SRR files

```sh
$ pwd
# output: ~/SRR13199080/
$ ls
# output: SRR13199080.sra
```

```sh
$ fasterq-dump SRR13199080.sra
# output fastq file size: 55.2 GB and 112 GB
# after gzip: 8.95 GB and 29.0 GB
# spots read      : 426,970,924
# reads read      : 853,941,848
# reads written   : 853,941,848

$ fasterq-dump SRR13199081.sra
# output fastq file size: 36.1 GB and 73.8G B
# after gzip: 5.86 GB and 19.8 GB
# spots read      : 280,795,719
# reads read      : 561,591,438
# reads written   : 561,591,438


$ fasterq-dump SRR13199085.sra
# output ffastq file size: 56.6GB and 115GB
# after gzip: 9.29 GB and 29.9 GB
# spots read      : 437,727,043
# reads read      : 875,454,086
# reads written   : 875,454,086

$ fasterq-dump SRR1319087.sra
# output ffastq file size: 50.6 GB and 103 GB
# after gzip: 8.07 GB and 26.6 GB
# spots read      : 392,020,289
# reads read      : 784,040,578
# reads written   : 784,040,578

$ fasterq-dump SRR13199088.sra
# output ffastq file size: 60.7 GB and 123 GB
# after gzip: 9.70 GB and 29.6 GB
# spots read      : 469,215,567
# reads read      : 938,431,134
# reads written   : 938,431,134

$ fasterq-dump SRR13199089.sra
# output ffastq file size: 81.9 GB and 166 GB
# after gzip: 
# spots read      : 631,622,354
# reads read      : 1,263,244,708
# reads written   : 1,263,244,708
```

# 4. CellRanger
## 4.1. Reference genome
GRCH38 2024 HUMAN from 10x genomics
```sh
# download reference genome file from 10X genomics
# ~10.6 GB
$ curl -O "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2024-A.tar.gz"

# unzip the file
# ~15.7 GB
$ tar -zxvf refdata-gex-GRCh38-2024-A.tar.gz
```

## 4.2. change file name

> normal-1_S1_L001_R1_001.fastq.gz\
> normal-1_S1_L001_R2_001.fastq.gz\
> normal-2_S1_L001_R1_001.fastq.gz\
> normal-2_S1_L001_R2_001.fastq.gz\
> normal-3_S1_L001_R1_001.fastq.gz\
> normal-3_S1_L001_R2_001.fastq.gz\
> hiv-1_S1_L001_R1_001.fastq.gz\
> hiv-1_S1_L001_R2_001.fastq.gz\
> hiv-2_S1_L001_R1_001.fastq.gz\
> hiv-2_S1_L001_R2_001.fastq.gz\
> hiv-3_S1_L001_R1_001.fastq.gz\
> hiv-3_S1_L001_R2_001.fastq.gz\

## 4.3. CellRanger
```sh
cellranger count --id=normal-1 \  # 생성되는 결과의 폴더명
    --fastqs=./SRR13199080 \     # fastq 파일이 있는 폴더의 위치
    --sample=normal-1 \            # metadata에 들어가는 sample 정보
    --transcriptome=./refdata-gex-GRCm39-2024-A \   # 참조 전사체의 위치
    --nosecondary \
    --create-bam=false

```
